development/studies-apex.md

```diff
@@ -0,0 +1,101 @@
+Structure of studies within Apex
+================================
+
+This page documents the implementation details of studies within Apex. Starting
+from the lowest layer.
+
+ManagedDirectory
+----------------
+
+This is a directory which is also a git repository. It provides a simplified
+interface to add, push, and pull files. It uses libgit2 as git
+implementation. All the expensive operations (push, pull, fetch, and checkout)
+are done on a separate thread, but are optionally synchronous through the use of
+different connection types.
+
+Study
+-----
+
+There are two types of study, public and private. A public study has got a
+single ManagedDirectory for the experiments. A private study has got an extra
+ManagedDirectory for the results.
+
+Furthermore it provides some convenience functions to easily query the
+experiments.
+
+StudyManager
+------------
+
+This class ties it all together and provides the interface for anything study
+related to the rest of Apex. It has several responsibilities.
+
+It has a global instance which should be used for all networking related
+functions[^1]. That is functions which update results or experiments. Note that
+networking related functions are not guaranteed to be reentrant depending on the
+transport protocol[^2]. Functions that only query the settings object (and the
+object's own data) are reentrant.
+
+In short, use the global instance from a single thread, it will automatically do
+the expensive operations on a different thread[^3].
+
+#### 1. Storing studies
+
+It uses QSettings to keep a list of linked studies and the current active
+study. It updates the settings each time this list gets altered.
+
+#### 2. Experiment life cycle
+
+It ties into the experiment life cycle as well. When a new experiment is started
+`StudyManager::newExperiment` should be called. This function will return an
+empty string if the experiment is part of the active study and if this study is
+private. This path should be set as the result file path as result collection
+depends on it. In any other case it will return an empty string and can be
+ignored.
+
+After the experiment `StudyManager::afterExperiment` should be called. This will
```

Comment:
which apex module would typically call these functions?

Reply:
afterExperiment is called by StudyModule (is explained in StudyModule section).

newExperiment is called by apexcontrol. Will add this to the doc.